### PR TITLE
Add TooManyPublicMethods rule

### PR DESCRIPTION
**Pull request by JeroenDeDauw**
Implement support for only checking the number of public methods as per #248

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
@ravage84 @manuelpichler Please comment and indicate which things should be done to make this mergeable. It appears to work, though I've only verified by means of unit tests so far.

Question: what do I do with the `author` and `copyright` tags? The new code is a modified copy of the old, so I can hardly just put my name in there, which I'm not sure would be the thing to do in this project to begin with.

Question: is this "duplication" fine? I suspect de-duplicating while keeping two rules will lead to a harder to maintain solution. And most duplication seems to be in the tests. Again, I'm thinking this is not an actual maintenance problem. Still, if two rules are not desired, having a single one will be easier to maintain. (Two rules are nicer in that you can have both with different limit and different ignore pattern.)

**Comment by ravage84** [comment]
About the ``author`` and ``copyright`` annotations in the file header doc block:
You can use the same copy of the file header doc block as in other files.
This is because the file header does not document the content of the file but is rather a "legal notice" included everywhere about the whole library.

In the past file header doc blocks __did__ document the content of a single file, but I think this view is outdated. It mostly dates back to the standard set by PEAR:
http://pear.php.net/manual/en/standards.header.php

**Comment by JeroenDeDauw** [comment]
In other words, the header doc is fine as I submitted it?

**Comment by JeroenDeDauw** [comment]
I'm still waiting on review here

**Comment by JeroenDeDauw** [comment]
Is there anything that is still expected from my side, or anything I can do to speed this up? I'd like to be able to use this rule out of the box in my projects...

File: src/site/rst/rules/codesize.rst [hunk 1]
```diff
@@ -223,14 +223,37 @@ TooManyMethods
...
+This rule has the following properties:
+
+=================================== =============== =======================================
+ Name                                Default Value   Description
+=================================== =============== =======================================
+ maxmethods                          25              The method count reporting threshold
+ ignorepattern                       (^(set|get))i   Ignore methods matching this regex
+=================================== =============== =======================================
+
+TooManyMethods
```
Comment (thiemowmde) [inline src/site/rst/rules/codesize.rst]: Whoops, TooManyPublicMethods please.

File: src/main/php/PHPMD/Rule/Design/TooManyMethods.php [hunk 2]
```diff
@@ -56,14 +56,6 @@
...
     /**
-     * the default regex pattern for ignore method names. override with
-     * "ignorepattern" rule setting.
-     *
-     * @var string
-     */
-    const DEFAULT_IGNORE_REGEXP = '(^(set|get))i';
-
-    /**
      * Regular expression that filter all methods that are ignored by this rule.
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/TooManyMethods.php]: "filters"

**Comment by ravage84** [comment]
@JeroenDeDauw ignore the file header doc block stuff.
I guess we should rework it anyway.

File: src/main/php/PHPMD/Rule/Design/TooManyPublicMethods.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+ * This rule class will detect all classes with too much public methods.
+ *
+ * @author    Manuel Pichler <mapi@phpmd.org>
+ * @copyright 2008-2014 Manuel Pichler. All rights reserved.
+ * @license   http://www.opensource.org/licenses/bsd-license.php BSD License
+ */
+class TooManyPublicMethods extends AbstractRule implements ClassAware
+{
+    /**
+     * Regular expression that filter all methods that are ignored by this rule.
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/TooManyPublicMethods.php]: "filters"

**Comment by ravage84** [comment]
I'm generelly OK with this PR.
What I don't like though, is that the new class is almost a 1:1 copy of the existing one with only 7 lines changed. But extending the new one from the old one does not improve it greatly anyway.

@manuelpichler can you make a decision please?

In any case, thanks for your work and especially our patience, @JeroenDeDauw .

**Comment by JeroenDeDauw** [comment]
I addressed the inline comments

**Comment by addshore** [comment]
@manuelpichler @ravage84 any progress here?

**Pull request merged**